common/achievement/bucket.ts

```diff
@@ -1,6 +1,94 @@
-import { BucketFormula } from './types';
+import moment from 'moment';
+import { Bucket, BucketConfig, BucketFormula, DefaultBucket, FilterBucket, GenericBucketConfig, TimeBucket } from './types';
+import { getRelationContext } from './util';
 
 type BucketCreatorDefs = Record<string, BucketFormula>;
 
 // Buckets are needed to pre-sort and aggregate certain events by types / a certain time window (e.g. weekly) etc.
-export const bucketCreatorDefs: BucketCreatorDefs = {};
+export const bucketCreatorDefs: BucketCreatorDefs = {
+    default: {
+        function: async (): Promise<GenericBucketConfig<DefaultBucket>> => {
+            return await { bucketKind: 'default', buckets: [] };
+        },
+    },
+    by_lecture_start: {
+        function: async (relation: string): Promise<GenericBucketConfig<TimeBucket>> => {
+            if (!relation) {
+                return { bucketKind: 'time', buckets: [] };
+            }
+            const context = await getRelationContext(relation);
+            if (!context[relation].lecture) {
+                return { bucketKind: 'time', buckets: [] };
+            }
+            return {
+                bucketKind: 'time',
+                buckets: context[relation].lecture.map((lecture) => ({
+                    kind: 'time',
+                    startTime: moment(lecture.start).subtract(10, 'minutes').toDate(),
+                    endTime: moment(lecture.start).add(lecture.duration, 'minutes').add(10, 'minutes').toDate(),
+                })),
+            };
+        },
+    },
+    by_weeks: {
+        function: async (): Promise<GenericBucketConfig<TimeBucket>> => {
+            // TODO - where did we get the number of weeks
```

Comment:
@dhenkel92 do you have a idea how we can get the number of weeks?

Reply:
what if we save the value of the last check as recordValue in the database and create 1 bucket more then last time?